agent/structs/check_type.go

```diff
@@ -44,8 +46,18 @@ type CheckType struct {
 type CheckTypes []*CheckType
 
 // Valid checks if the CheckType is valid
-func (c *CheckType) Valid() bool {
-	return c.IsTTL() || c.IsMonitor() || c.IsHTTP() || c.IsTCP() || c.IsDocker()
+func (c *CheckType) Valid() (bool, error) {
+	valid := c.IsTTL() || c.IsMonitor() || c.IsHTTP() || c.IsTCP() || c.IsDocker()
+	var err error
+	if !valid {
+		err = fmt.Errorf(c.validateMsg())
```

Comment:
The `validateMsg` helper is odd here and kind of confusing how it works. This seems like it should be more explicit about checking the fields here and produce the errors directly from in here. There are some edge cases where this was sloppy where they can give a TTL and an interval, so that's weird as well. I'd change this to something like

```
func (c *CheckType) Validate() error {
    if c.IsScript() || c.HTTP != "" || c.TCP != "" {
        if c.Interval <= 0 {
            return fmt.Errorf("Interval must be >0 for script, HTTP, or TCP checks")
        }
    
        if c.TTL > 0 {
            return fmt.Errorf("Interval and TTL cannot both be specified")
        }
    } else if c.TTL <= 0 {
        return fmt.Errorf("TTL must be >0 for TTL checks")
    }

    return nil
}
```

Reply:
I find that a bit hard to follow. How about:

```go
func (c *CheckType) Validate() error {
	intervalCheck := c.IsScript() || c.HTTP != "" || c.TCP != ""

	if c.Interval > 0 && c.TTL > 0 {
		return fmt.Errorf("Interval and TTL cannot both be specified")
	}
	if intervalCheck && c.Interval <= 0 {
		return fmt.Errorf("Interval must be > 0 for Script, HTTP, or TCP checks")
	}
	if !intervalCheck && c.TTL <= 0 {
		return fmt.Errorf("TTL must be > 0 for TTL checks")
	}
	return nil
}
```

Reply:
@magiconair's is cleaner than my suggestion